### PR TITLE
(RK-307) Ignore branches for deployment with specified prefixes

### PR DESCRIPTION
**Pull request by andersonmills**
This is a partial implementation of feature request #574

This PR implements ignoring branches based on a prefix list provided in the r10k.yaml configuration file for a given source. There is a new setting option for a source called ignore_branch_prefixes:. If specified, then only branches that do not start with a string in the list will be deployed into environments. If ignore_branch_prefixes: is not specified, then all branches will be deployed. This makes it a backwards compatible change.

Please see the updated documentation for configuration in this PR for examples.

Comments and reviews appreciated.

The original work for this came from @nmaludy and is adapted from #757

### CODE REVIEW

**Review by nmaludy — APPROVED** [review]

File: doc/dynamic-environments/configuration.mkd
Comment (haus) [inline doc/dynamic-environments/configuration.mkd]: You shouldn't need 2 groups here. `'^(prod|qa)$'` should also work?

File: doc/dynamic-environments/configuration.mkd
Comment (haus) [inline doc/dynamic-environments/configuration.mkd]: You shouldn't need a grouping here at all?

File: doc/dynamic-environments/configuration.mkd
Comment (haus) [inline doc/dynamic-environments/configuration.mkd]: This can be `'^(master|production)$'`.

File: doc/dynamic-environments/configuration.mkd
Comment (haus) [inline doc/dynamic-environments/configuration.mkd]: Excluding branches this way feels a bit awkward. I wonder if having two filters, one for include and one for exclude would be better? That way you could write positive regexes in both cases, which is usually easier. I think in that case you would add `'^dev-.*'` to an exclude regex and any matches would be excluded. Not sure if that is better or not.

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: This regex match syntax is fairly weird for ruby. I would expect something more like:

```
filter = Regexp.new(@branch_filter)
...
result = filter.match(branch)
...
```

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: The double negation here is funky. I would recommend a partially positive test instead such as...

`if @branch_filter && !@branch_filter.empty?`
or
`if @branch_filter && @branch_filter.size > 0`

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: The select call could be lifted into its own method and then unit tested on its own. `filter_branches` or something of the sort.

**Comment by nmaludy** [comment]
@andersonmills not sure how you want me to proceed with fixing the changes that @haus requested?

File: doc/dynamic-environments/configuration.mkd
Comment (nmaludy) [reply doc/dynamic-environments/configuration.mkd]: What happens if they specify both filters, then which filter applies?

I would be more inclined to implement an additional option branch_filter_type that can either be include or exclude. Then the regex will be interpreted as either inclusive or exclusive.

File: doc/dynamic-environments/configuration.mkd
Comment (nmaludy) [reply doc/dynamic-environments/configuration.mkd]: Yes, but it also doesn't matter. We're not actually using the group information so having 1 or 2 groups makes no difference.

File: doc/dynamic-environments/configuration.mkd
Comment (nmaludy) [reply doc/dynamic-environments/configuration.mkd]: Yes, but it also doesn't matter. We're not actually using the group information so having 1 or 2 groups makes no difference.

File: doc/dynamic-environments/configuration.mkd
Comment (nmaludy) [reply doc/dynamic-environments/configuration.mkd]: No, it doesn't matter in terms of functionality, but does help to logically separate out the prefix from the "extra stuff at the end".

**Comment by andersonmills** [comment]
@nmaludy It was my intention to make the changes. :) Please feel free to comment on @haus 's comments, though.

File: doc/dynamic-environments/configuration.mkd
Comment (andersonmills) [reply doc/dynamic-environments/configuration.mkd]: I'm a big fan of the simplicity of having one filter and using the regex negation. I expect that users getting to this level of usage will be able to follow the examples or be able to look it up quickly.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: Making the change.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: Opting for the former suggestion.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: Workin' on that.

File: doc/dynamic-environments/configuration.mkd
Comment (haus) [reply doc/dynamic-environments/configuration.mkd]: I feel like adding a setting that is a regex is probably a sign that we're taking the wrong approach to solving the problem. Do users actually need full regex matching, or would a simple allow/reject on a prefix or lists of allowed/denied environments be adequate?
Regexes tend to be simple in implementation, but not in actual usage, and debugging them can be challenging.

Comment (andersonmills) [reply doc/dynamic-environments/configuration.mkd]: I can see users wanting to both be able to use a list of denied/allowed environments, and a list of denied/allowed prefixes. As I start thinking about how to document how this works for users, it gets ...uh... complicated. Do you have an configuration spec in mind?

Comment (haus) [reply doc/dynamic-environments/configuration.mkd]: I would probably start with something like `ignore_branches` or `ignore_branch_prefixes` as a yaml array. I suspect that having a way to reject environments with a given prefix would be sufficient for most users.

Comment (haus) [reply doc/dynamic-environments/configuration.mkd]: `ignore_branch_prefixes` could also be a list of environments to ignore.

Comment (andersonmills) [reply doc/dynamic-environments/configuration.mkd]: I like it.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: Well, I lifted it out, but it's a private method, as is the branch_names method. Should we be unit testing the private methods? I can, but I thought that was considered bad testing methodology.

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: i think if you use `reject` (https://ruby-doc.org/core-2.4.2/Enumerable.html#method-i-reject) instead of `select` here you can avoid the negation of `result` below.

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: What is the `.?` doing in the regex? I assume it is an optional `any character`, but I'm not sure why it is needed.

File: lib/r10k/source/svn.rb
Comment (haus) [inline lib/r10k/source/svn.rb]: I think the same change as was made to git could be applied here. (Both the conditional change and the ignore_branch_prefix change).

**Comment by haus** [comment]
@andersonmills i don't have a great answer about testing private methods. I think testing the method seems nice because it is small and contained, but I think in that case it probably shouldn't be a private method? No strong feelings.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: Changing.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: It is not. Removing.

File: lib/r10k/source/svn.rb
Comment (andersonmills) [reply lib/r10k/source/svn.rb]: D'oh! It's been so long since I paid attention to svn in r10k. Adding.

**Comment by andersonmills** [comment]
@haus I made those methods public and wrote tests. Those methods are now tested individually, but with the same fail cases as the previous larger scale tests (which aren't really unit tests, anyway).

File: lib/r10k/source/git.rb
Comment (haus) [inline lib/r10k/source/git.rb]: This looks pretty good. I think `Regexp.new("^#{Regexp.union(ignore_prefixes)}")` is equivalent, but I'm not sure it's any cleaner.

File: spec/unit/source/git_spec.rb
Comment (haus) [inline spec/unit/source/git_spec.rb]: I think this test description is a little funky. When i first read it I presumes it would be a test without using the ignore_branch_prefixes setting. Maybe `describe "generate_environments respects ignore_branch_prefixes setting"`?

File: spec/unit/source/git_spec.rb
Comment (haus) [inline spec/unit/source/git_spec.rb]: I think a test branch with dev or test at the end or the middle of a valid branch name would be good.

File: spec/unit/source/svn_spec.rb
Comment (haus) [inline spec/unit/source/svn_spec.rb]: This describe should probably be the same as whatever you end up with for git.

File: spec/unit/source/svn_spec.rb
Comment (haus) [inline spec/unit/source/svn_spec.rb]: Same comment as in the git about having a test case with dev/test in the middle or end of a valid branch name.

**Comment by haus** [comment]
This looks great. I had some feedback for the test cases, but I think this is good to merge after that?

File: spec/unit/source/git_spec.rb
Comment (andersonmills) [reply spec/unit/source/git_spec.rb]: Yeah, this has been through a few iterations to get here.

Comment (andersonmills) [reply spec/unit/source/git_spec.rb]: Yeah this went through too many editing iterations.

File: lib/r10k/source/git.rb
Comment (andersonmills) [reply lib/r10k/source/git.rb]: It's maybe a little clearer (maybe).

File: spec/unit/source/git_spec.rb
Comment (andersonmills) [reply spec/unit/source/git_spec.rb]: Good idea.

Comment (andersonmills) [reply spec/unit/source/git_spec.rb]: Adding this caused me to notice that "dev-test" is not a valid environment name, anyway.

**Comment by andersonmills** [comment]
@haus I think this hits the test changes you wanted. :)

**Review by haus — APPROVED** [review]
Sweet. Looks great.

**Comment by andersonmills** [comment]
Rebased after a merge up. But it seems like I shouldn't see all that, then. Hm. Noting that all the merged up commits have checks beside them already.

Aha! It looks like there is a problem caused by rebasing with commits shared between two users. I though it's going to be messy, I don't think there's a way around it.

Merging.

**Pull request merged**
